### ISSUE TITLE
Infinite loop when direct deserializing a List

### ISSUE THREAD

**Issue by FYGonzalo** [jefffhaynes/BinarySerializer#61]
Hello again Jeff. I am the same that gave you headaches on #52 but I promise this time will be better (I swear I readed a C# book from the start to the end).

I have a "little complex" process of serialization/deserialization. Basically what I need is, for deserialization process, deserializate the "PacketHeader" (contains general information about the packet structure) and based on its properties, I must, if the checksum of the buffer is equal to the PacketHeader checksum property, decrypt nor decompress the rest of the stream before continue the deserialization process.

So this are my classes:

```csharp
public class Packet : IBinarySerializable
    {
        [FieldOrder(0)]
        public PacketHeader Header { get; set; }
        [FieldOrder(1)]
        [FieldLength("Header.Length")]
        [FieldCount("Header.OperationsCount")]
        public List<Message> Messages { get; set; }

        public void Deserialize(Stream stream, Endianness endianness, BinarySerializationContext serializationContext)
        {
            BinarySerializer serializer = new BinarySerializer();

            // This will be part of PacketHeader custom serialization
            PacketHeaderEncryption.Decrypt(stream);
            stream.Seek(0, SeekOrigin.Begin);
            // End

            Header = (PacketHeader)serializer.Deserialize(stream, typeof(PacketHeader));

            if (Header.Status.HasFlag(BufferStatus.Encrypted))
            {

            }

            if (Header.Status.HasFlag(BufferStatus.Compressed))
            {

            }

            Messages = (List<Message>)serializer.Deserialize(stream, typeof(List<Message>));
        }

        public void Serialize(Stream stream, Endianness endianness, BinarySerializationContext serializationContext)
        {
            throw new NotImplementedException();
        }     
    }

public class PacketHeader 
    {
        [FieldOrder(0)]
        public ushort Length { get; set; }
        [FieldOrder(1)]
        public short OperationsCount { get; set; }
        [FieldOrder(2)]
        public BufferStatus Status { get; set; }
        [FieldOrder(3)]
        public byte Checksum { get; set; }       
    }

public class Message
    {
        [FieldOrder(0)]
        public ushort Length { get; set; }
        [FieldOrder(1)]
        public short Code { get; set; }
        [FieldOrder(2)]
        [FieldLength("Length", ConverterType = typeof(MessageLengthConverter))]
        public byte[] Buffer { get; set; }
    }
```

When I try to deserialize the List of Message inside the Deserialize method of Packet the serializer never ends. I subscribed to debug events of binaryserializer and saw that it deserializates well the first and unique object (Message) in the stream and then infinitely deserializes again the Length of the same message.

Basically what I have in the stream after deserialize the PacketHeader is just one or more Message serialized
{[0x0008]([0x0002][0x01, 0x01, 0x01, 0x01, 0x01, 0x01])} + more Message
_Note: I use a converter for length because the Length property of Message is Code length + Buffer length_

Is the problem because I am manually deserializing the List and so it doesn't get the bindings (Count and Length)?

Maybe I should pass those values into context parameter of Deserialize method, but I don't know how to do that, and even not sure if that is a way to go.

Thank you again and thank for great framework!

**Comment by jefffhaynes** [comment]
Sorry I've been traveling and haven't had a chance to look at this yet but I'll ty to get to it this weekend.

**Comment by jefffhaynes** [comment]
I haven't been able to figure this out yet but maybe try setting the Length of an object that contains the code and buffer and get rid of the converter.  For example, you would be better off defining:

```c#
public class MessagePayload
{
    [FieldOrder(0)]
    public short Code { get; set; }

    [FieldOrder(1)]
    public byte[] Buffer { get; set; }
}

public class Message
{
    [FieldOrder(0)]
    public int Length { get; set; }

    [FieldOrder(1)]
    [FieldLength("Length")]
    public MessagePayload Payload { get; set; }
}
```

The framework knows that short is two bytes long so it will figure out the balance of the object (Buffer) correctly.

**Comment by FYGonzalo** [comment]
I bet that my serialized stream (not serialized with this framework) were broken.

Sorry again.

**Comment by FYGonzalo** [comment]
I don't wanna fill your github with nonsense questions so I will do another in this issue.

Is there any problem with specifing FieldLenght for byte arrays? I am using the lastest version from nuget (7.2.0) and when the field has less bytes than FieldLength specifies, at serialization, there is not padding.
Also, when there are more bytes than FieldLength specifies, the serialization throws an exception about the length of the stream.

The documentation says it should work:
> Field lengths can apply to anything that is sizable including strings, **arrays**, lists, streams, and even objects.

Tested with strings and works..

**Comment by jefffhaynes** [comment]
Are you passing a constant length to the FieldLength constructor?  If you are binding, the source of the binding will be updated to match the length of the byte array. If you pass a constant value then the framework will pad during serialization.

---------------
I much prefer the sharpest criticism of a single intelligent man to the thoughtless approval of the masses - Johannes Kepler

________________________________

**Comment by FYGonzalo** [comment]
Its a constant length. I have tried on a new project with this class:

```csharp
class TestingClass
    {
        [FieldOrder(0)]
        [FieldLength(2)]
        public byte[] Field1 { get; set; }

        [FieldOrder(1)]
        [FieldLength(6)]
        public byte[] Field2 { get; set; }
    }
```

```csharp
 static void Main(string[] args)
        {

            TestingClass testingClass = new TestingClass()
            {
                Field1 = new byte[] { 0x01 },
                Field2 = new byte[] { 0x99, 0x98, 0x97, 0x96, 0x95, 0x94}
            };

            byte[] serializedObject;

            using (MemoryStream stream = new MemoryStream())
            {
                BinarySerializer serializer = new BinarySerializer();
                serializer.Serialize(stream, testingClass);
                serializedObject = stream.ToArray();
            }

        }
```

And the result is a byte array with:
> { 0x01, 0x99, 0x98, 0x97, 0x96, 0x95, 0x94 }

Also, If I assign the byte array { 0x01, 0x02, 0x03 } to Field1, when serializing this exception is raised:
> Unable to write beyond end of stream limit.

If Field1 is null then the serializated stream only has 6 bytes.

**Comment by FYGonzalo** [comment]
Another problem* that I found happens when using FieldAlignment on a class, and then use that class as property of another one.
_*Maybe the framework is working as expected and I missed something, but I really tried hard to find whats going on and couldn't found anything_

I am having inconsistency with the value used in FieldAlignment.

This is the parent class:
```csharp
public class Parent
    {
        [FieldOrder(0)]
        public int Value { get; set; }

        [FieldOrder(1)]
        public Entry Entry { get; set; }
    }
```

And this is the class used as property (taken from documentation)
```csharp
public class Entry
    {
        [FieldOrder(0)]
        public byte Length { get; set; }

        [FieldOrder(1)]
        [FieldAlignment(4, FieldAlignmentMode.LeftOnly)]
        [FieldLength("Length")]
        public string Value { get; set; }
    }
```

And this is what I am doing to test:
```csharp
byte[] serializedClass = new byte[]
            {
                0x05, 0x05, 0x05, 0x05,
                0x08, 0x70, 0x70, 0x70,
                0x64, 0x64, 0x64, 0x64,
                0x65, 0x65, 0x63, 0x63,
                0x63, 0x63, 0x63, 0x63,
                0x63, 0x63, 0x63, 0x63
            };

BinarySerializer serializer = new BinarySerializer();
Parent parent = serializer.Deserialize<Parent>(serializedClass);
```
0x63 = c
0x64 = d
0x65 = e
0x70 = p

In this case, parent object ends with:
```xml
<parent>
    <Value>0x05050505</Value>
    <Entry>
        <Length>8</Length>
        <Value>ddddeecc</Value>
    </Entry>
</parent>
```

But when I set the FieldAlignment of Entry to any number bigger than 4, the offset is considered from the start of the entire serialized array.
ie with FieldAlignment = 5 the result is:
```xml
<parent>
    <Value>0x05050505</Value>
    <Entry>
        <Length>8</Length>
        <Value>pppdddde</Value>
    </Entry>
</parent>
```

with FieldAlignment = 8 the result is the same as when its 4:
```xml
<parent>
    <Value>0x05050505</Value>
    <Entry>
        <Length>8</Length>
        <Value>ddddeecc</Value>
    </Entry>
</parent>
```

**Comment by jefffhaynes** [comment]
Const length issue moved to #63

**Comment by jefffhaynes** [comment]
Field alignment issue moved to #64

**Comment by jefffhaynes** [comment]
Thanks, these were both bugs.  The field alignment one actually uncovered some other issues.  Sorry it took me a while to get to them but they should be fixed in 7.3.
